sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala

```diff
@@ -176,34 +176,46 @@ class WholeStageCodegenSuite extends SparkPlanTest with SharedSQLContext {
     })
 
     assert(wholeStageCodeGenExec.isDefined)
-    wholeStageCodeGenExec.get.asInstanceOf[WholeStageCodegenExec].doCodeGen()._1
+    wholeStageCodeGenExec.get.asInstanceOf[WholeStageCodegenExec].doCodeGen()
   }
 
   test("SPARK-21603 check there is a too long generated function") {
     withSQLConf(SQLConf.WHOLESTAGE_MAX_LINES_PER_FUNCTION.key -> "1500") {
-      val ctx = genGroupByCodeGenContext(30)
+      val (ctx, _) = genGroupByCodeGenContext(30)
       assert(ctx.isTooLongGeneratedFunction === true)
     }
   }
 
   test("SPARK-21603 check there is not a too long generated function") {
     withSQLConf(SQLConf.WHOLESTAGE_MAX_LINES_PER_FUNCTION.key -> "1500") {
-      val ctx = genGroupByCodeGenContext(1)
+      val (ctx, _) = genGroupByCodeGenContext(1)
       assert(ctx.isTooLongGeneratedFunction === false)
     }
   }
 
   test("SPARK-21603 check there is not a too long generated function when threshold is Int.Max") {
     withSQLConf(SQLConf.WHOLESTAGE_MAX_LINES_PER_FUNCTION.key -> Int.MaxValue.toString) {
-      val ctx = genGroupByCodeGenContext(30)
+      val (ctx, _) = genGroupByCodeGenContext(30)
       assert(ctx.isTooLongGeneratedFunction === false)
     }
   }
 
   test("SPARK-21603 check there is a too long generated function when threshold is 0") {
     withSQLConf(SQLConf.WHOLESTAGE_MAX_LINES_PER_FUNCTION.key -> "0") {
-      val ctx = genGroupByCodeGenContext(1)
+      val (ctx, _) = genGroupByCodeGenContext(1)
       assert(ctx.isTooLongGeneratedFunction === true)
     }
   }
+
+  test("SPARK-21871 turn off whole-stage codegen if bytecode size goes over HugeMethodLimit") {
+    withSQLConf(SQLConf.WHOLESTAGE_MAX_LINES_PER_FUNCTION.key -> Int.MaxValue.toString) {
+      val (_, code) = genGroupByCodeGenContext(20)
+      val errMsg = intercept[IllegalArgumentException] {
+        CodeGenerator.compile(code)
+      }.getMessage
+      assert(errMsg.contains("the size of GeneratedClass.agg_doAggregateWithKeys is 9182 and " +
```

Comment:
The hard-coded method name and size is a bit too strict. Can we relax that a little bit so that it's more resilient to the naming and the size of the generated classes/methods?